tiny_dnn/core/framework/tensor.h

```diff
@@ -61,45 +63,173 @@
 
 namespace tiny_dnn {
 
+template<typename U = float_t>
+class TensorStorage {
+ public:
+    /**
+     * Initializes an empty tensor storage.
+     * @return
+     */
+    TensorStorage() {
+    }
+
+    /**
+     * Constructor that assepts a vector of shape and create a TensorStorage
+     * with a size equivalent to that shape.
+     * @param shape array containing N integers, sizes of dimensions
+     * @return
+     */
+    explicit TensorStorage(const std::vector<size_t> &shape) {
+        resize(shape);
+    }
+
+    /**
+     * Constructor that assepts an initializer list  of shape and create a
+     * TensorStorage with a size equivalent to that shape.
+     * @param shape array containing N integers, sizes of dimensions
+     * @return
+     */
+    explicit TensorStorage(std::initializer_list<size_t> const &shape) {
+        std::vector<size_t> tmp(shape.size());
+        std::copy(shape.begin(), shape.end(), tmp.begin());
+        resize(tmp);
+    }
+
+    void sync() {
+        if (data_is_on_host_)
+            toDevice();
+        else
+            fromDevice();
+    }
+
+    typename std::vector<U, aligned_allocator < U, 64>>::
+    iterator host_data(size_t offset) {
+        return host_data_.begin() + offset;
+    }
+
+    typename std::vector<U, aligned_allocator < U, 64>>::
+    const_iterator host_data(size_t offset) const {
+        return host_data_.begin() + offset;
+    }
+
+    void resize(const std::vector<size_t> &sz) {
+        host_data_->resize(std::accumulate(std::begin(sz),
+                                           std::end(sz),
+                                           size_t(1),
+                                           std::multiplies<size_t>()), U(0));
+        //TODO(Randl): device data
```

Comment:
vector::resize with zero initialization is fine for most of the uses.
But please consider adding another resize method that uses below vector method to resize and initialize at once efficiently.
```
template <class InputIterator>
void assign (InputIterator first, InputIterator last);
```

and what about below methods?
```
begin
end
cbegin
cend
```